R/combineWidgets.R

```diff
@@ -0,0 +1,186 @@
+#'   to fit the number of rows.
+#' @param colsize
+#'   Same as rowsize but for the height of the columns of the layout.
+#' @param byrow
+#'   If \code{TRUE}, then the layout is filled by row. Else it is filled by
+#'   column.
+#' @param titleCSS
+#'   A character containing css properties to modify the appearance of the title
+#'   of the view.
+#' @param width
+#'   Total width of the layout (optional, defaults to automatic sizing)
+#' @param height
+#'   Total height of the layout (optional, defaults to automatic sizing)
+#' @return
+#'   A HTML widget.
+#'
+#' @details
+#' The function only allows table like layout : each row has the same number of
+#' columns and reciprocally. But it is possible to create more complex layout
+#' by nesting combined htmlwidgets. (see examples)
+#'
+#' @examples
+#' if (require(plotly)) {
+#'   data(iris)
+#'
+#'   combineWidgets(title = "The Iris dataset",
+#'     plot_ly(iris, x = ~Sepal.Length, type = "histogram", nbinsx = 20),
+#'     plot_ly(iris, x = ~Sepal.Width, type = "histogram", nbinsx = 20),
+#'     plot_ly(iris, x = ~Petal.Length, type = "histogram", nbinsx = 20),
+#'     plot_ly(iris, x = ~Petal.Width, type = "histogram", nbinsx = 20)
+#'   )
+#'
+#'   # Create a more complex layout by nesting combinedWidgets
+#'   combineWidgets(title = "The iris data set: sepals", ncol = 2, colsize = c(2,1),
+#'     plot_ly(iris, x = ~Sepal.Length, y = ~Sepal.Width, type = "scatter",
+#'             mode = "markers", color = ~Species),
+#'     combineWidgets(
+#'       plot_ly(iris, x = ~Sepal.Length, type = "histogram", nbinsx = 20),
+#'       plot_ly(iris, x = ~Sepal.Width, type = "histogram", nbinsx = 20)
+#'     )
+#'   )
+#' }
+#'
+#' @export
+combineWidgets <- function(..., nrow = NULL, ncol = NULL, title = NULL,
+                           rowsize = 1, colsize = 1, byrow = TRUE,
+                           titleCSS = "",
+                           width = NULL, height = NULL) {
+  widgets <- lapply(list(...), function(x) {
+    if (is.null(x$preRenderHook)) return(x)
+    x$preRenderHook(x)
+  })
+  nwidgets <- length(widgets)
+
+  # Get number of rows and cols
+  if (!is.null(nrow) && !is.null(ncol) && nrow * ncol < nwidgets) {
+    stop("There are too much widgets compared to the number of rows and columns")
+  } else if (is.null(nrow) && !is.null(ncol)) {
+    nrow <- ceiling(nwidgets / ncol)
+  } else if (!is.null(nrow) && is.null(ncol)) {
+    ncol <- ceiling(nwidgets / nrow)
+  } else if (is.null(nrow) && is.null(ncol)) {
+    nrow <- ceiling(sqrt(nwidgets))
+    ncol <- ceiling(nwidgets / nrow)
+  }
+
+  ncells <- nrow * ncol
+
+  # Relative size of rows and cols
+  rowsize <- rep(rowsize, length.out=nrow)
+  colsize <- rep(colsize, length.out = ncol)
+
+  # Get the html ID of each widget
+  elementId <- sapply(widgets[1:ncells], function(x) {
+    if (is.null(x)) res <- NULL
+    else res <- x$elementId
+
+    if (is.null(res)) res <- paste0("widget", floor(runif(1, max = 1e9)))

```

Comment:
should probably use `htmlwidgets::createWidgetId` [line](https://github.com/ramnathv/htmlwidgets/blob/master/R/widgetid.R#L24) but it unfortunately is not exported